Avoid namespace collisions when collapsing graph to custom node

mjkkirschner opened the pull request:
### Purpose

Collapsing a graph to a custom node can create errors if there is a namespace conflict in types used in the generated input nodes.

Adds a new visitor to the shortQualifiedNameReplacer and uses it to replace names when creating custom node symbols when collapsing.

Will cherry pick this to 2.1 branch if no issues found.

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@aparajit-pratap 


### FYIs

@smangarole

aparajit-pratap commented on src/DynamoCore/Core/CustomNodeManager.cs:
```diff
@@ -674,7 +678,8 @@ internal bool TryGetInfoFromPath(string path, bool isTestMode, out CustomNodeInf
                 nodeGraph.ElementResolver,
```
I see the nodeGraph is instantiated from an xml file? Is that why it doesn't get the correct element resolver as we are now using JSON?

mjkkirschner replied:
here is where we use the deserialized element resolver when loading a JSON graph and pass it to the home workspace or customNode workspace that was just deserialized.

https://github.com/DynamoDS/Dynamo/blob/ec10f936824152e7dd7d6d019efdcda0d78a5264/src/DynamoCore/Graph/Workspaces/SerializationConverters.cs#L532

mjkkirschner commented:
@aparajit-pratap I've updated this with one potential solution.

aparajit-pratap commented on src/DynamoCore/Core/CustomNodeManager.cs:
```diff
@@ -1090,10 +1103,35 @@ internal bool TryGetNodeInfo(string name, out CustomNodeInfo info)
+                                {
+                                    node.InputSymbol += ":" + typedParameter.Type.Name;
+                                    this.AsLogger().LogError($"{e.Message}: could not generate a short type name for {typedParameter.Type.Name}");
+                                }
```
Does this set `InputSymbol` to the full name even though it is unable to find a short name? If not, it should.

mjkkirschner replied:
yep, thats what it does.

aparajit-pratap commented on src/DynamoCore/Engine/NodeToCode/NodeToCode.cs:
```diff
@@ -320,6 +321,55 @@ public ShortestQualifiedNameReplacer(ClassTable classTable, ElementResolver reso
+                return node;
+            
+            // build an idlistnode from the full type name
+            var identListNode = CoreUtils.CreateNodeFromString(node.datatype.Name);
```
We can just pass `node.TypeAlias` to  `CreateNodeFromString`.

mjkkirschner replied:
will try that - thanks.

aparajit-pratap commented on src/DynamoCore/Engine/NodeToCode/NodeToCode.cs:
```diff
@@ -320,6 +321,55 @@ public ShortestQualifiedNameReplacer(ClassTable classTable, ElementResolver reso
+            // build an idlistnode from the full type name
+            var identListNode = CoreUtils.CreateNodeFromString(node.datatype.Name);
+            if (identListNode == null)
+                return null;
```
It might be safer to return `node` here just in case it's still valid.

aparajit-pratap commented on src/DynamoCore/Engine/NodeToCode/NodeToCode.cs:
```diff
@@ -320,6 +321,55 @@ public ShortestQualifiedNameReplacer(ClassTable classTable, ElementResolver reso
+            // visit the idlist 
+            var shortNameNode = identListNode.Accept(this);
+            if(shortNameNode == null)
+                return null;
```
It might be safer to return node here just in case it's still valid.

aparajit-pratap commented on src/DynamoCore/Engine/NodeToCode/NodeToCode.cs:
```diff
@@ -320,6 +321,55 @@ public ShortestQualifiedNameReplacer(ClassTable classTable, ElementResolver reso
+            {
+                var idNode = (shortNameNode as IdentifierNode);
+                resultingNode.TypeAlias = idNode.Value;
+            }
```
I think we can just do: `resultingNode.TypeAlias = shortNameNode.ToString();`

aparajit-pratap commented on src/DynamoCore/Engine/NodeToCode/NodeToCode.cs:
```diff
@@ -320,6 +321,55 @@ public ShortestQualifiedNameReplacer(ClassTable classTable, ElementResolver reso
+            
+
+            var resultingNode = new TypedIdentifierNode();
+            resultingNode.Name = node.Name;
```
```
var resultingNode = new TypedIdentifierNode{
Name = node.Name,
Value = node.Name,
...
}

mjkkirschner replied:
👍

aparajit-pratap commented on src/DynamoCore/Engine/NodeToCode/NodeToCode.cs:
```diff
@@ -320,6 +321,55 @@ public ShortestQualifiedNameReplacer(ClassTable classTable, ElementResolver reso
+            node.Name = resultingNode.Name;
+            node.Value = resultingNode.Value;
+            node.datatype = resultingNode.datatype;
+            node.TypeAlias = resultingNode.TypeAlias;
```
Does `node` have to be passed as a `ref` parameter in order to directly assign it to `resultingNode`? Is that why you have modified its properties individually here?

mjkkirschner replied:
yep, I found that assigning `node = resultantNode` did not update the node I passed in, and using a ref parameter requires touching the ASTreplacer base class - so this is not ideal, but it does seem to work.

aparajit-pratap commented on src/DynamoCore/Core/CustomNodeManager.cs:
```diff
@@ -1077,10 +1083,17 @@ internal bool TryGetNodeInfo(string name, out CustomNodeInfo info)
                             parameters = funcDesc.Parameters.ToList();
 
+                            // if the node is an instance member the function won't contain a 
+                            // parameter for this type so we need to geneate a new typedParameter.
```
typo: generate

aparajit-pratap commented:
@mjkkirschner I think this looks good overall. Couple of other comments you could address here. I will look at the test cases you add before we decide to merge this. Thanks for sticking with this and fixing it :)

mjkkirschner commented:
@aparajit-pratap I've moved the `ShortQualifiedNameReplacer` to it's own file but kept it in dynamo core because it was used by public methods in dynamoCore (ie `ReplaceWithShorestQualifiedName`) and I wanted to keep the underlying class internal. 

I've added a few tests for shortNameReplacer that test the `typedIdentifer` visitor and some customNodeCollapse tests that use FFITarget to create conflicts.

I have yet to run all the tests on self serve CI.

aparajit-pratap commented on test/DynamoCoreTests/CustomNodes.cs:
```diff
@@ -893,6 +893,85 @@ public void TestCustomNodeFromCollapsedNodeHasTypes()
+            var inputs = CurrentDynamoModel.CurrentWorkspace.Nodes.OfType<Symbol>().ToList();
+            Assert.IsNotNull(inputs);
+
+            var pointParams = inputs;
```
where is `pointParams` used?

aparajit-pratap commented on test/DynamoCoreTests/CustomNodes.cs:
```diff
@@ -893,6 +893,85 @@ public void TestCustomNodeFromCollapsedNodeHasTypes()
+
+            Assert.IsTrue(inputs.All(t =>
+            {
+                Console.WriteLine(t.InputSymbol);
```
What's the `Console.WriteLine` for?

aparajit-pratap commented on test/DynamoCoreTests/CustomNodes.cs:
```diff
@@ -893,6 +893,85 @@ public void TestCustomNodeFromCollapsedNodeHasTypes()
+            var inputs = CurrentDynamoModel.CurrentWorkspace.Nodes.OfType<Symbol>().ToList();
+            Assert.IsNotNull(inputs);
+
+            var pointParams = inputs;
```
same here.

aparajit-pratap commented:
@mjkkirschner excellent! Some minor comments and then LGTM!
